### PR TITLE
Model limitations

### PR DESCRIPTION
**Pull request by hflynn**
Tidying and updating https://www.openmicroscopy.org/site/support/omero5.1/developers/Model.html while adding limitations as per https://trello.com/c/uV9FuKGO/341-5-1-limitations

cc @mtbc 

Needs - text reviewing for WHOLE page, can we add any items from the To Do list? Are there items there that should have a card? Can we add API doc links anywhere/would it be helpful to do so?

### CODE REVIEW

**Comment by mtbc** [comment]
I'll certainly read this page over this week with a view to larger-scale improvements as well as the smaller changes, but @joshmoore or @jburel you too may wish to weigh in before @hflynn works any more on the page next week.

File: omero/developers/Model.txt [hunk 1]
```diff
@@ -385,19 +382,19 @@ Each instance can be found in one of several states. Quickly, they are:
...
 
 Identity, references, and versions
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Critical for understanding these states is understanding the concepts of
 identity and versioning as it relates to ORM. Every object has an id
-field that if created by the backend will not be :literal:`null`. However, every
-table does not have a primary key field -- subclasses contain a foreign
+field that if created by the backend will not be :literal:`null`. However,
```
Comment (mtbc) [inline omero/developers/Model.txt]: "... that is created ..."

File: omero/developers/Model.txt [hunk 2]
```diff
@@ -653,29 +649,29 @@ It is possible to both add and remove a link directly:
...
 
-        // someDataset is now included in someProject
+    // someDataset is now included in someProject
 
-        update.deleteObject(link);
-        // or update.deleteObject(new ProjectDatasetLink(link.getId(), false)); // a proxy
+    update.deleteObject(link);
+    // or update.deleteObject(new ProjectDatasetLink(link.getId(), false)); // a proxy
+
+    // Now they the Dataset is not included,
```
Comment (mtbc) [inline omero/developers/Model.txt]: remove "they"

File: omero/developers/Model.txt [hunk 3]
```diff
@@ -683,11 +679,11 @@ equal. That is:
...
 
-        someProject.linkDataset( someDataset );
-        updatedProject = update.saveAndReturnObject( someProject );
+    someProject.linkDataset( someDataset );
+    updatedProject = update.saveAndReturnObject( someProject );
 
-        updatedProject.unlinkDataset( someDataset );
-        update.saveObject( updateProject ); // will no __nothing__ !
+    updatedProject.unlinkDataset( someDataset );
+    update.saveObject( updateProject ); // will no __nothing__ !
```
Comment (mtbc) [inline omero/developers/Model.txt]: "will *do* ..."

**Comment by mtbc** [comment]
Regarding `developer-model-pdi.png` now a `Channel` too may be annotated.

**Comment by mtbc** [comment]
There's a ```
file:`model/target/generated/src` directory
``` that's missing its initial `:`.

**Comment by mtbc** [comment]
"takes the generate ome.model classes as input" → "... generated ..."

**Comment by mtbc** [comment]
too many " in, `External to the server code is the "blitz"" layer.`

File: omero/developers/Model.txt [hunk 4]
```diff
@@ -129,17 +129,16 @@ include:
 -  details
 -  … any SQL keyword
 
-
 Properties
-~~~~~~~~~~
-Mutable, annotated, global.
+^^^^^^^^^^
 
+Mutable, annotated, global.
```
Comment (mtbc) [inline omero/developers/Model.txt]: @joshmoore: Is this a to-do for expanding, or does it somehow continue the "not allowed" thread above?

**Comment by mtbc** [comment]
"polymorophism" → "polymorphism"

**Comment by mtbc** [comment]
@joshmoore: In the discussion of CGLIB and Javassist requiring use of `Utils#trueClass()`, would it make sense to also mention https://docs.jboss.org/hibernate/orm/3.5/javadocs/org/hibernate/Hibernate.html#getClass%28java.lang.Object%29 ?

**Comment by mtbc** [comment]
```
``LazyIniitalizationExceptions``
```
should be
```
``LazyInitializationException``s
```

File: omero/developers/Model.txt [hunk 5]
```diff
@@ -426,7 +423,7 @@ example:
 
 ::
 
-       select p from Project p left outer join fetch p.datasetLinks as links left outer join fetch links.child as dataset"
+    select p from Project p left outer join fetch p.datasetLinks as links left outer join fetch links.child as dataset"
```
Comment (mtbc) [inline omero/developers/Model.txt]: also remove the final "

**Comment by mtbc** [comment]
For the rest of the page, I could generate minor suggestions, but I am not sure if it would be worth it or, if it were, if it would be easier if I just opened a subsequent PR rather than explaining.

Larger-scale ideas if we want to trim the page down:
* Move all the earlier stuff about the type language and code generation to a different page. Most readers will never need to know, they'll just need the later bit about working with the model.
* The entity lifecycle and other Hibernate introduction is great, but perhaps it is possible to simply refer the reader to others' Hibernate documentation / tutorial.
* Move the future topics section to a Trello card.

**Comment by mtbc** [comment]
I should add, some of the future topics, like versioning, do already seem to be at least somewhat partly addressed in the text.

File: omero/developers/Model.txt (hunk 1 above)
Comment (hflynn) [inline omero/developers/Model.txt]: is it not intended to be "that, if created by the backend, will not be…"? Changing it to 'is' stops the 'will not be' bit making sense.

**Comment by hflynn** [comment]
Ok, I've tackled all the minor formatting and typos. I'm happy for this to be merged tomorrow so Mark can open a new PR to do further work this week while I'm on leave.
I agree that if there is a sizeable chunk we could split off into a separate page that would be good, this is very long at the moment! 
Re: Trello card - It would be good if these could be reviewed rather than copied into a card en masse.

**Comment by mtbc** [comment]
```
/home/travis/build/openmicroscopy/ome-documentation/omero/developers/Model.txt:423: WARNING: Inline literal start-string without end-string.
```

I think I'll have to rely on @joshmoore to do that review, I'm not too sure what some of the items actually involve.

File: omero/developers/Model.txt (hunk 1 above)
Comment (mtbc) [inline omero/developers/Model.txt]: Doh, of course you're correct.

File: omero/developers/Model.txt [hunk 6]
```diff
@@ -567,7 +563,7 @@ multiple threads, but simply due to stale state.
 
 .. note:: 
     There is an issue with multiple users in which a 
-    ``SecurityViolation`` is thrown instead of an ``OptimisticLockException``. 
+    ``SecurityViolation`` is thrown instead of an ``OptimisticLockException``.
     See :ticket:`1649` for more information.
```
Comment (mtbc) [inline omero/developers/Model.txt]: remove mentions of this ticket, see http://trac.openmicroscopy.org.uk/ome/ticket/1649#comment:13

File: omero/developers/Model.txt [hunk 7]
```diff
@@ -546,16 +542,16 @@ not know whether the object should be removed or not.
...
+    Dataset dataset = iQuery.findByQuery("select d from Dataset d "+
+            "join fetch d.projectsLinks links "+
+            "join fetch links.parent "+
+            "where d.id = :id", new Parameters().addId(5L));
+    Project newProject = dataset.linkedProjectList().get(0);
+    assert newProject.getId().equals(oldProject.getId()) : "same object";
+    assert newProject.sizeOfDatasetLinks() == oldProject.sizeOfDatasetLinks() :
+            "if this is false, then saving oldProject is a problem";
 
 Without optimistic locks, return oldProject, trying to save oldProject
```
Comment (mtbc) [inline omero/developers/Model.txt]: I am not quite sure what this "return oldProject" part is about.

File: omero/developers/Model.txt (hunk 4 above)
Comment (joshmoore) [inline omero/developers/Model.txt]: It's for explanding. These are interfaces which need explaining. I'd think remove for the moment, and we can add back as necessary.

File: omero/developers/Model.txt (hunk 7 above)
Comment (joshmoore) [inline omero/developers/Model.txt]: Should be removed.

**Comment by hflynn** [comment]
@joshmoore given the java issue and release probably being moved forward, are you likely to have time for further review of this page or shall I just merge it now?

**Comment by joshmoore** [comment]
I'd say let's get it in and then we can review at our leisure. (If we decide we have leisure this week, so be it)

**Pull request merged**